ndc-models/src/lib.rs

```diff
@@ -420,9 +420,37 @@ pub struct Query {
     pub offset: Option<u32>,
     pub order_by: Option<OrderBy>,
     pub predicate: Option<Expression>,
+    ///
+    pub groups: Option<Grouping>,
 }
 // ANCHOR_END: Query
 
+// ANCHOR: Grouping
+#[skip_serializing_none]
+#[derive(Clone, Debug, PartialEq, Serialize, Deserialize, JsonSchema)]
+#[schemars(title = "Grouping")]
+pub struct Grouping {
+    /// Dimensions along which to partition the data
+    pub dimensions: IndexMap<String, Dimension>,
+    /// Fields to fetch in each group
+    pub fields: IndexMap<String, Field>,
```

Comment:
I've dropped this, but remind me why it was removed? I think there's a good argument for a drill-down operation to be supported easily by adding a field to your GraphQL query. 

And it would be more reliable to fetch rows per-group rather than to expect the engine to get the grouping right.

A related possible feature is to express relationships from dimension columns. For example, if I group articles by `author_id`, it's sensible to want to be able to fetch the related author, once-per-group and not once-per-row.

Reply:
The argument was that if you cared about row data, you could just pull back the rows and group yourself client side. There's very little advantage having the server do the grouping for you, since you'd be pulling all row data anyway.

Aggregates over groups are different as they don't require pulling row data to the client, and can be optimised by the server.

Reply:
I don't think it makes sense to have the engine try to replicate the behavior of the database. If we group by a `citext` column for example, is the engine going to do the right thing? Also, that implies that the semantics of grouping are fixed, but maybe they shouldn't be. Sorted grouping and unsorted grouping are both valid IMO, for example.

Finally, there maybe be a future concrete benefit to doing this server side: as I say, you can fetch less data if you can express relationships from dimension columns.

Reply:
> I don't think it makes sense to have the engine try to replicate the behavior of the database

Nobody has suggested the v3-engine should be doing this. The GraphQL API user would do it.

> you can fetch less data if you can express relationships from dimension columns.

Yeah, that's a valid point; or at least you can not do two queries on the GraphQL API client end. One where you query the dimensions and one where you perform another query to load the related entities. To clarify, are you saying then that fields are only useful for dimension columns, or do you still think loading group row data is necessary. They seem separate to me.

Is there any reason we couldn't add this functionality incrementally in the future if customers asked for it, rather than add it now?

Reply:
To clarify, I see this as useful:

```graphql
query MyQuery {
  Invoice_groups(grouping_keys: {Customer: {_scalar_field: CustomerId}}) {
    group_key {
      Customer {
        CustomerId # This is the actual group key
        Address { # Load the related address!
          City
          Country
          PostalCode
          State
          StreetAddress
        }
      }
    }
  }
}
```

but it doesn't translate to SQL very cleanly, because technically all the related Address fields need to be added to the grouping key for this to work (and it would work right now, if you did add them, no?).

This is not so useful, because you end up just getting the Customer Address just repeated over and over for each group row, rather than once per customer group:

```graphql
query MyQuery {
  Invoice_groups(grouping_keys: {Customer: {_scalar_field: CustomerId}}) {
    group_key {
      Customer {
        CustomerId # This is the actual group key
      }
    }
    group_rows { # This is not in the RFC
      Customer {
        Address { # Load the related address!
          City
          Country
          PostalCode
          State
          StreetAddress
        }
      }
   }
  }
}
```

Reply:
Actually, looking at the type of `Dimension` again, we actually can't express the first graphql query at all, since we can't address object-related columns. Perhaps this needs to change and then we could achieve what you're talking about like this:

```
query MyQuery {
  Invoice_groups(
    grouping_keys: [ 
      {Customer: {_scalar_field: CustomerId}},
      {Customer: {Address: {_scalar_field: StreetAddress}}},
      {Customer: {Address: {_scalar_field: City}}},
      {Customer: {Address: {_scalar_field: Country}}},
      {Customer: {Address: {_scalar_field: PostalCode}}},
      {Customer: {Address: {_scalar_field: State}}}
    ]
  ) {
    group_key {
      Customer {
        CustomerId
        Address {
          City
          Country
          PostalCode
          State
          StreetAddress
        }
      }
    }
  }
}
```

Reply:
> Is there any reason we couldn't add this functionality incrementally in the future if customers asked for it, rather than add it now?

Yes, we can add it later, I just wanted to make sure I understood the decision.